charger/easee.go

```diff
@@ -306,6 +317,46 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
 	case easee.DYNAMIC_CHARGER_CURRENT:
 		c.dynamicChargerCurrent = value.(float64)
 	case easee.CHARGER_OP_MODE:
+		//New charging session pending, reset internal value of SESSION_ENERGY to 0, and its observation timestamp to "now".
+		//This should be done in a proper way by the api, but it's not.
+		//Remember value of LIFETIME_ENERGY as start value of the charging session
+		if easee.ModeDisconnected >= c.opMode && easee.ModeAwaitingStart <= value.(int) {
+			c.sessionEnergy = 0
+			loc, _ := time.LoadLocation("UTC")
+			c.obsTime[easee.SESSION_ENERGY] = time.Now().In(loc)
+			c.sessionStartLifetimeEnergy = easee.NEED_SESSION_START_ENERGY
+		}
+		//OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
+		if easee.ModeCharging == value.(int) && easee.ModeCharging != c.opMode {
+			c.ticker = time.NewTicker(5 * time.Minute)
+			c.stopTicker = make(chan bool)
+			go func() {
+				for {
+					select {
+						case <-c.stopTicker:
+							return
+						case t := <-c.ticker.C:
+							c.RequestLifetimeEnergyUpdate()
+					}
+				}
+			}()
+		}
+
+		//OpMode changed FROM charging to something else - stop ticker
+		if easee.ModeCharging == c.opMode && easee.ModeCharging != value.(int) {
+			c.ticker.Stop()
```

Comment:
Der ticker sollte nur in der Goroutine leben

Reply:
So, ich hab das probiert, ich kenne nicht die Geheimnisse von Go, daher wider besserer Befürchtung wie folgt getestet:
 
```
	case easee.CHARGER_OP_MODE:
	  
	  ...
	  
		var ticker *time.Ticker
		var stopTicker chan struct{}

		// OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
		if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {
			ticker = time.NewTicker(5 * time.Minute)
			stopTicker = make(chan struct{})
			go func() {
				for {
					select {
					case <-stopTicker:
						return
					case <-ticker.C:
						c.requestLifetimeEnergyUpdate()
					}
				}
			}()
		}

		// OpMode changed FROM charging to something else - stop ticker
		if c.opMode == easee.ModeCharging && value.(int) != easee.ModeCharging {
// ================
			ticker.Stop()
			stopTicker <- struct{}{}
		}
```

Irgendwie erwartungsgemäß knallt das bei // ================, da ticker beim Aufruf von 
`func (c *Easee) ProductUpdate(i json.RawMessage) { ` 
im `case easee.CHARGER_OP_MODE` und wahrer Bedingung `if c.opMode == easee.ModeCharging && value.(int) != easee.ModeCharging {`   nil ist. 
Wird ja auch nur beim Aufruf der Funktion im Fall `if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {` gefüllt.

`[easee ] TRACE 2023/09/25 11:26:57 level=info connection=lEwEPYssHz2GwWWHgEcZBA event=panic in target method error=runtime error: invalid memory address or nil pointer dereference name=ProductUpdate reaction=send completion with error`

Übersehe ich hier irgendein Konstrukt in Go, das es in C oder Java nicht gibt, um hier den Ticker nur in der Goroutine leben zu lassen..?

Ich habe das also erstmal so gelassen. Der ganze Rest ist aber umgesetzt :)

Reply:
. oder ist das `ticker.Stop()` völlig unnötig, und `stopTicker <- struct{}{}` ausreichend um den Ticker zu stoppen?
Dann könnte der `ticker `in der Tat wie im Test lokal definiert sein, und nur `stopTicker` global (oder wie auch immer es in Go denn heißt.. ;))

Reply:
Ja, so sollte das sein. Wenn Du aus der go routine via return aussteigst, geht der Ticker out-of-scope und wird dann irgendwann von der GC eingesammelt. Du kannst aber auch trotzdem den Ticker noch innerhalb der Go Routine direkt vor dem return anhalten. Vielleicht unnötig - da bin auch ich jetzt überfragt.